File: src/cmds/files.ts
import * as fs from 'fs';
import * as path from 'path';
import { run } from '../core/runner';
import { stripAnsi, truncate, readFileIfExists, safeJsonParse } from '../core/utils';
import { TokConfig } from '../core/config';
import { FilterLevel, filterCode, langFromPath, smartSummary, extractStructure } from '../core/filter';
import { HandlerResult } from './git';

export function handleLs(args: string[], ultra: boolean, config: TokConfig): HandlerResult {
  const target = args.find((a) => !a.startsWith('-')) || '.';
  const start = Date.now();
  let filtered = '';
  let raw = '';
  let exitCode = 0;
  try {
    raw = readDirRaw(target);
    filtered = formatTree(target, ultra, config);
  } catch (err) {
    raw = String(err);
    filtered = raw;
    exitCode = 1;
  }
  return {
    filteredOutput: filtered,
    exitCode,
    rawOutput: raw,
    cmdType: 'ls',
    execMs: Date.now() - start,
  };
}

function readDirRaw(target: string): string {
  try {
    const entries = fs.readdirSync(target);
    return entries.join('\n');
  } catch (err) {
    return String(err);
  }
}

export function formatTree(dirPath: string, ultra: boolean, config: TokConfig): string {
  const noise = new Set(config.noiseDirectories);
  const maxDepth = config.filters.ls.maxDepth;

  let totalFiles = 0;
  let totalDirs = 0;

  type Node = { name: string; isDir: boolean; children: Node[]; fileCount: number };
  function walk(p: string, depth: number): Node {
    const node: Node = {
      name: path.basename(p) || p,
      isDir: true,
      children: [],
      fileCount: 0,
    };
    if (depth > maxDepth) return node;
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(p, { withFileTypes: true });
    } catch {
      return node;
    }
    const files = entries.filter((e) => e.isFile());
    const dirs = entries.filter((e) => e.isDirectory() && !noise.has(e.name));
    node.fileCount = files.length;

    for (const f of files) totalFiles++;
    for (const d of dirs) {
      totalDirs++;
      node.children.push(walk(path.join(p, d.name), depth + 1));
    }
    if (files.length <= 5) {
      for (const f of files) {
        node.children.push({ name: f.name, isDir: false, children: [], fileCount: 0 });
      }
    }
    return node;
  }

  const root = walk(dirPath, 0);

  if (ultra) {
    const items: string[] = [];
    let extra = 0;
    for (const c of root.children) {
      if (c.isDir) {
        if (items.length < 6) items.push(`${c.name}/${c.fileCount}`);
        else extra++;
      } else if (items.length < 6) {
        items.push(c.name);
      }
    }
    if (extra > 0) items.push(`[+${extra} dirs]`);
    return items.join(' ');
  }

  const lines: string[] = [];
  lines.push(`${root.name === '.' ? path.resolve(dirPath) : root.name}/`);
  renderTree(root, '', lines);
  lines.push('');
  lines.push(`Total: ${totalFiles} files in ${totalDirs} directories`);
  return lines.join('\n');
}

function renderTree(node: { name: string; isDir: boolean; children: { name: string; isDir: boolean; children: any[]; fileCount: number }[]; fileCount: number }, prefix: string, lines: string[]): void {
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    const last = i === node.children.length - 1;
    const branch = last ? '└─' : '├─';
    const nextPrefix = prefix + (last ? '   ' : '│  ');
    if (child.isDir) {
      if (child.fileCount > 5) {
        lines.push(`${prefix}${branch} ${child.name}/ (${child.fileCount} files)`);
      } else {
        lines.push(`${prefix}${branch} ${child.name}/`);
        renderTree(child as any, nextPrefix, lines);
      }
    } else {
      lines.push(`${prefix}${branch} ${child.name}`);
    }
  }
}

export function handleCat(args: string[], ultra: boolean, config: TokConfig): HandlerResult {
  const file = args.find((a) => !a.startsWith('-'));
  const start = Date.now();
  if (!file) {
    return {
      filteredOutput: 'usage: tok cat <file>',
      exitCode: 2,
      rawOutput: '',
      cmdType: 'cat',
      execMs: Date.now() - start,
    };
  }
  const raw = readFileIfExists(file) || '';
  if (!raw) {
    return {
      filteredOutput: `cannot read: ${file}`,
      exitCode: 1,
      rawOutput: '',
      cmdType: 'cat',
      execMs: Date.now() - start,
    };
  }
  const lang = langFromPath(file);
  const level = pickLevel(args, config.filters.cat.defaultLevel, ultra);
  let filtered = filterCode(raw, lang, level);
  filtered = truncate(filtered, config.filters.cat.maxLines);
  return {
    filteredOutput: filtered,
    exitCode: 0,
    rawOutput: raw,
    cmdType: 'cat',
    execMs: Date.now() - start,
  };
}

function pickLevel(args: string[], def: FilterLevel, ultra: boolean): FilterLevel {
  if (ultra) return FilterLevel.Aggressive;
  if (args.includes('--aggressive')) return FilterLevel.Aggressive;
  if (args.includes('--minimal')) return FilterLevel.Minimal;
  if (args.includes('--none')) return FilterLevel.None;
  return def;
}

export function handleSmart(args: string[], _ultra: boolean): HandlerResult {
  const file = args.find((a) => !a.startsWith('-'));
  const start = Date.now();
  if (!file) {
    return {
      filteredOutput: 'usage: tok smart <file>',
      exitCode: 2,
      rawOutput: '',
      cmdType: 'smart',
      execMs: Date.now() - start,
    };
  }
  const raw = readFileIfExists(file) || '';
  if (!raw) {
    return {
      filteredOutput: `cannot read: ${file}`,
      exitCode: 1,
      rawOutput: '',
      cmdType: 'smart',
      execMs: Date.now() - start,
    };
  }
  const lang = langFromPath(file);
  return {
    filteredOutput: smartSummary(raw, lang),
    exitCode: 0,
    rawOutput: raw,
    cmdType: 'smart',
    execMs: Date.now() - start,
  };
}

export function handleGrep(args: string[], ultra: boolean, config: TokConfig): HandlerResult {
  const cmd = which('rg') ? 'rg' : 'grep';
  let useArgs = args.slice();

  // Detect single-file targets; force -H so grep prefixes the filename even with one file,
  // otherwise the grouping regex (which expects file:line:content) sees only line:content
  // and reports "0 matches".
  const positional = useArgs.filter((a) => !a.startsWith('-'));
  const looksLikeSingleFile = positional.length >= 2 && positional.slice(1).filter((p) => existsSyncSafe(p)).length === 1;

  if (cmd === 'grep') {
    if (!useArgs.includes('-rn') && !useArgs.includes('-n')) useArgs = ['-rn', ...useArgs];
    if (looksLikeSingleFile && !useArgs.includes('-H')) useArgs = ['-H', ...useArgs];
  } else if (cmd === 'rg') {
    // ripgrep already prefixes filenames per match by default — no change needed.
  }

  const result = run(cmd, useArgs);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    filtered = groupByFile(raw, ultra, config.filters.grep.maxMatches);
  } catch {
    filtered = raw;
  }
  return {
    filteredOutput: filtered || (result.exitCode === 0 ? '0 matches' : raw),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType: 'grep',
    execMs: result.execMs,
  };
}

function existsSyncSafe(p: string): boolean {
  try { return fs.existsSync(p); } catch { return false; }
}

export function groupByFile(raw: string, ultra: boolean, maxMatches: number): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => l.trim());
  const byFile = new Map<string, number[]>();
  let total = 0;
  for (const line of lines) {
    const m = /^(.+?):(\d+):/.exec(line);
    if (!m) continue;
    total++;
    const arr = byFile.get(m[1]) || [];
    arr.push(parseInt(m[2], 10));
    byFile.set(m[1], arr);
  }
  if (total === 0) return ultra ? '0' : '0 matches';

  if (ultra) {
    const parts: string[] = [];
    for (const [file, ls] of byFile) {
      const short = file.split(/[\/\\]/).pop() || file;
      parts.push(`${short}:${ls.length}`);
    }
    return parts.slice(0, 8).join(' ');
  }

  const out: string[] = [];
  let shown = 0;
  for (const [file, ls] of byFile) {
    if (shown + ls.length > maxMatches) {
      out.push(`[+${total - shown} more matches]`);
      break;
    }
    const lines = ls.slice(0, 10).map((n) => `L${n}`).join(', ');
    const more = ls.length > 10 ? `, +${ls.length - 10} more` : '';
    out.push(`${file}: ${ls.length} match${ls.length === 1 ? '' : 'es'} (${lines}${more})`);
    shown += ls.length;
  }
  return out.join('\n');
}

function which(cmd: string): boolean {
  try {
    const result = run(process.platform === 'win32' ? 'where' : 'which', [cmd]);
    return result.exitCode === 0;
  } catch {
    return false;
  }
}

export function handleFind(args: string[], ultra: boolean): HandlerResult {
  const start = Date.now();

  // On Windows the system `find` is a substring search tool with completely different
  // arguments from unix find, and `where` only finds executables on PATH. Implement
  // our own portable walker so `tok find <dir> -name "*.ts"` works the same everywhere.
  let lines: string[];
  let exitCode = 0;
  let raw = '';
  if (process.platform === 'win32' || hasUnsupportedFindArg(args)) {
    const { dir, pattern } = parseFindArgs(args);
    try {
      lines = walkFiles(dir, pattern);
      raw = lines.join('\n');
    } catch (err) {
      lines = [];
      raw = String(err);
      exitCode = 1;
    }
  } else {
    const result = run('find', args);
    raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
    exitCode = result.exitCode;
    lines = stripAnsi(raw).split('\n').filter((l) => l.trim());
  }

  let filtered: string;
  if (ultra) {
    filtered = `${lines.length} files`;
  } else if (lines.length > 50) {
    filtered = `${lines.length} files found\n${lines.slice(0, 50).join('\n')}\n[+${lines.length - 50} more]`;
  } else {
    filtered = lines.join('\n') || '0 files';
  }
  return {
    filteredOutput: filtered,
    exitCode,
    rawOutput: raw,
    cmdType: 'find',
    execMs: Date.now() - start,
  };
}

function hasUnsupportedFindArg(args: string[]): boolean {
  // Args we know our portable walker doesn't implement; fall back to system find on POSIX.
  const supported = new Set(['-name', '-iname']);
  return args.some((a) => a.startsWith('-') && !supported.has(a));
}

function parseFindArgs(args: string[]): { dir: string; pattern: RegExp | null } {
  let dir = '.';
  let pattern: RegExp | null = null;
  let caseInsensitive = false;
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === '-name' || a === '-iname') {
      caseInsensitive = a === '-iname';
      const glob = args[++i] || '';
      pattern = globToRegex(glob, caseInsensitive);
    } else if (!a.startsWith('-') && i === 0) {
      dir = a;
    } else if (!a.startsWith('-')) {
      // Additional positional — usually a path filter; ignore for portability.
    }
  }
  return { dir, pattern };
}

function globToRegex(glob: string, ci: boolean): RegExp {
  let re = '';
  for (const ch of glob) {
    if (ch === '*') re += '.*';
    else if (ch === '?') re += '.';
    else if ('.+^${}()|[]\\'.includes(ch)) re += '\\' + ch;
    else re += ch;
  }
  return new RegExp('^' + re + '$', ci ? 'i' : '');
}

function walkFiles(dir: string, pattern: RegExp | null): string[] {
  const results: string[] = [];
  const queue: string[] = [dir];
  while (queue.length) {
    const cur = queue.shift()!;
    let entries: fs.Dirent[];
    try { entries = fs.readdirSync(cur, { withFileTypes: true }); } catch { continue; }
    for (const e of entries) {
      const full = path.join(cur, e.name);
      if (e.isDirectory()) {
        if (e.name === 'node_modules' || e.name === '.git') continue;
        queue.push(full);
      } else if (e.isFile()) {
        if (!pattern || pattern.test(e.name)) results.push(full);
      }
    }
  }
  return results;
}

export function handleDiff(args: string[], ultra: boolean): HandlerResult {
  const result = run('diff', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  const clean = stripAnsi(raw);
  const lines = clean.split('\n');
  let added = 0;
  let removed = 0;
  for (const line of lines) {
    if (line.startsWith('>') || (line.startsWith('+') && !line.startsWith('+++'))) added++;
    else if (line.startsWith('<') || (line.startsWith('-') && !line.startsWith('---'))) removed++;
  }
  const filtered = ultra ? `+${added}-${removed}` : `+${added} -${removed} (${lines.length} diff lines)`;
  return {
    filteredOutput: filtered,
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType: 'diff',
    execMs: result.execMs,
  };
}

export function handleJson(args: string[], _ultra: boolean): HandlerResult {
  const file = args.find((a) => !a.startsWith('-'));
  const start = Date.now();
  if (!file) {
    return {
      filteredOutput: 'usage: tok json <file>',
      exitCode: 2,
      rawOutput: '',
      cmdType: 'json',
      execMs: Date.now() - start,
    };
  }
  const raw = readFileIfExists(file) || '';
  if (!raw) {
    return {
      filteredOutput: `cannot read: ${file}`,
      exitCode: 1,
      rawOutput: '',
      cmdType: 'json',
      execMs: Date.now() - start,
    };
  }
  const parsed = safeJsonParse(raw);
  if (parsed === null) {
    return {
      filteredOutput: `invalid JSON: ${file}`,
      exitCode: 1,
      rawOutput: raw,
      cmdType: 'json',
      execMs: Date.now() - start,
    };
  }
  const structure = extractStructure(parsed);
  return {
    filteredOutput: JSON.stringify(structure, null, 2),
    exitCode: 0,
    rawOutput: raw,
    cmdType: 'json',
    execMs: Date.now() - start,
  };
}

File: internal/handlers/files.go
package handlers

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"slices"
	"strings"
	"time"

	"github.com/Kevalgor12/tok-proxy/internal/config"
	"github.com/Kevalgor12/tok-proxy/internal/filter"
	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

func since(t time.Time) int { return int(time.Since(t).Milliseconds()) }

func firstPositional(args []string, def string) string {
	for _, a := range args {
		if !strings.HasPrefix(a, "-") {
			return a
		}
	}
	return def
}

// ---- ls --------------------------------------------------------------------

func HandleLs(args []string, ultra bool, cfg config.Config) Result {
	target := firstPositional(args, ".")
	start := time.Now()
	return Result{
		Filtered: formatTree(target, ultra, cfg),
		Exit:     0,
		Raw:      readDirRaw(target),
		CmdType:  "ls",
		ExecMs:   since(start),
	}
}

func readDirRaw(target string) string {
	entries, err := os.ReadDir(target)
	if err != nil {
		return err.Error()
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		names = append(names, e.Name())
	}
	return strings.Join(names, "\n")
}

type lsNode struct {
	name      string
	isDir     bool
	fileCount int
	children  []*lsNode
}

func formatTree(dirPath string, ultra bool, cfg config.Config) string {
	noise := map[string]bool{}
	for _, d := range cfg.NoiseDirectories {
		noise[d] = true
	}
	maxDepth := cfg.Filters.Ls.MaxDepth
	totalFiles, totalDirs := 0, 0

	var walk func(p string, depth int) *lsNode
	walk = func(p string, depth int) *lsNode {
		name := filepath.Base(p)
		if name == "" {
			name = p
		}
		node := &lsNode{name: name, isDir: true}
		if depth > maxDepth {
			return node
		}
		entries, err := os.ReadDir(p)
		if err != nil {
			return node
		}
		var files, dirs []os.DirEntry
		for _, e := range entries {
			switch {
			case e.IsDir():
				if !noise[e.Name()] {
					dirs = append(dirs, e)
				}
			case e.Type().IsRegular():
				files = append(files, e)
			}
		}
		node.fileCount = len(files)
		totalFiles += len(files)
		for _, d := range dirs {
			totalDirs++
			node.children = append(node.children, walk(filepath.Join(p, d.Name()), depth+1))
		}
		if len(files) <= 5 {
			for _, f := range files {
				node.children = append(node.children, &lsNode{name: f.Name()})
			}
		}
		return node
	}
	root := walk(dirPath, 0)

	if ultra {
		var items []string
		extra := 0
		for _, c := range root.children {
			if c.isDir {
				if len(items) < 6 {
					items = append(items, fmt.Sprintf("%s/%d", c.name, c.fileCount))
				} else {
					extra++
				}
			} else if len(items) < 6 {
				items = append(items, c.name)
			}
		}
		if extra > 0 {
			items = append(items, fmt.Sprintf("[+%d dirs]", extra))
		}
		return strings.Join(items, " ")
	}

	rootName := root.name
	if rootName == "." {
		if abs, err := filepath.Abs(dirPath); err == nil {
			rootName = abs
		}
	}
	lines := []string{rootName + "/"}
	renderTree(root, "", &lines)
	lines = append(lines, "", fmt.Sprintf("Total: %d files in %d directories", totalFiles, totalDirs))
	return strings.Join(lines, "\n")
}

func renderTree(node *lsNode, prefix string, lines *[]string) {
	for i, child := range node.children {
		last := i == len(node.children)-1
		branch, nextPrefix := "├─", prefix+"│  "
		if last {
			branch, nextPrefix = "└─", prefix+"   "
		}
		switch {
		case child.isDir && child.fileCount > 5:
			*lines = append(*lines, fmt.Sprintf("%s%s %s/ (%d files)", prefix, branch, child.name, child.fileCount))
		case child.isDir:
			*lines = append(*lines, fmt.Sprintf("%s%s %s/", prefix, branch, child.name))
			renderTree(child, nextPrefix, lines)
		default:
			*lines = append(*lines, fmt.Sprintf("%s%s %s", prefix, branch, child.name))
		}
	}
}

// ---- cat / smart / json ----------------------------------------------------

func HandleCat(args []string, ultra bool, cfg config.Config) Result {
	file := firstPositional(args, "")
	start := time.Now()
	if file == "" {
		return Result{Filtered: "usage: tok cat <file>", Exit: 2, CmdType: "cat", ExecMs: since(start)}
	}
	raw, ok := util.ReadFileIfExists(file)
	if !ok || raw == "" {
		return Result{Filtered: "cannot read: " + file, Exit: 1, CmdType: "cat", ExecMs: since(start)}
	}
	level := pickLevel(args, cfg.Filters.Cat.DefaultLevel, ultra)
	filtered := util.Truncate(filter.FilterCode(raw, filter.LangFromPath(file), level), cfg.Filters.Cat.MaxLines)
	return Result{Filtered: filtered, Exit: 0, Raw: raw, CmdType: "cat", ExecMs: since(start)}
}

func pickLevel(args []string, def config.FilterLevel, ultra bool) config.FilterLevel {
	switch {
	case ultra, slices.Contains(args, "--aggressive"):
		return config.FilterAggressive
	case slices.Contains(args, "--minimal"):
		return config.FilterMinimal
	case slices.Contains(args, "--none"):
		return config.FilterNone
	default:
		return def
	}
}

func HandleSmart(args []string, _ bool) Result {
	file := firstPositional(args, "")
	start := time.Now()
	if file == "" {
		return Result{Filtered: "usage: tok smart <file>", Exit: 2, CmdType: "smart", ExecMs: since(start)}
	}
	raw, ok := util.ReadFileIfExists(file)
	if !ok || raw == "" {
		return Result{Filtered: "cannot read: " + file, Exit: 1, CmdType: "smart", ExecMs: since(start)}
	}
	return Result{Filtered: filter.SmartSummary(raw, filter.LangFromPath(file)), Exit: 0, Raw: raw, CmdType: "smart", ExecMs: since(start)}
}

func HandleJson(args []string, _ bool) Result {
	file := firstPositional(args, "")
	start := time.Now()
	if file == "" {
		return Result{Filtered: "usage: tok json <file>", Exit: 2, CmdType: "json", ExecMs: since(start)}
	}
	raw, ok := util.ReadFileIfExists(file)
	if !ok || raw == "" {
		return Result{Filtered: "cannot read: " + file, Exit: 1, CmdType: "json", ExecMs: since(start)}
	}
	var parsed any
	if json.Unmarshal([]byte(raw), &parsed) != nil {
		return Result{Filtered: "invalid JSON: " + file, Exit: 1, Raw: raw, CmdType: "json", ExecMs: since(start)}
	}
	b, err := json.MarshalIndent(filter.ExtractStructure(parsed), "", "  ")
	if err != nil {
		return Result{Filtered: "invalid JSON: " + file, Exit: 1, Raw: raw, CmdType: "json", ExecMs: since(start)}
	}
	return Result{Filtered: string(b), Exit: 0, Raw: raw, CmdType: "json", ExecMs: since(start)}
}

// ---- grep ------------------------------------------------------------------

func HandleGrep(args []string, ultra bool, cfg config.Config) Result {
	cmd := "grep"
	if _, err := exec.LookPath("rg"); err == nil {
		cmd = "rg"
	}
	useArgs := append([]string{}, args...)

	// grep needs -H to print the filename even for a single-file target; without it the
	// "file:line:content" grouping regex only sees "line:content" and reports 0 matches.
	var positionals []string
	for _, a := range args {
		if !strings.HasPrefix(a, "-") {
			positionals = append(positionals, a)
		}
	}
	singleFileTargets := 0
	if len(positionals) >= 2 {
		for _, p := range positionals[1:] {
			if fileExists(p) {
				singleFileTargets++
			}
		}
	}
	looksLikeSingleFile := len(positionals) >= 2 && singleFileTargets == 1

	if cmd == "grep" {
		if !slices.Contains(useArgs, "-rn") && !slices.Contains(useArgs, "-n") {
			useArgs = append([]string{"-rn"}, useArgs...)
		}
		if looksLikeSingleFile && !slices.Contains(useArgs, "-H") {
			useArgs = append([]string{"-H"}, useArgs...)
		}
	}

	r := run.Run(cmd, useArgs)
	raw := combined(r)
	filtered := groupByFile(raw, ultra, cfg.Filters.Grep.MaxMatches)
	if filtered == "" {
		if r.ExitCode == 0 {
			filtered = "0 matches"
		} else {
			filtered = raw
		}
	}
	return Result{Filtered: filtered, Exit: r.ExitCode, Raw: raw, CmdType: "grep", ExecMs: r.ExecMs}
}

var grepLineRe = regexp.MustCompile(`^(.+?):(\d+):`)

func groupByFile(raw string, ultra bool, maxMatches int) string {
	clean := util.StripAnsi(raw)
	var fileOrder []string
	byFile := map[string][]int{}
	total := 0
	for _, line := range strings.Split(clean, "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		m := grepLineRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		total++
		if _, ok := byFile[m[1]]; !ok {
			fileOrder = append(fileOrder, m[1])
		}
		byFile[m[1]] = append(byFile[m[1]], atoi(m[2]))
	}
	if total == 0 {
		if ultra {
			return "0"
		}
		return "0 matches"
	}

	if ultra {
		var parts []string
		for _, file := range fileOrder {
			short := file
			if i := strings.LastIndexAny(file, `/\`); i >= 0 {
				short = file[i+1:]
			}
			parts = append(parts, fmt.Sprintf("%s:%d", short, len(byFile[file])))
		}
		return strings.Join(limit(parts, 8), " ")
	}

	var out []string
	shown := 0
	for _, file := range fileOrder {
		ls := byFile[file]
		if shown+len(ls) > maxMatches {
			out = append(out, fmt.Sprintf("[+%d more matches]", total-shown))
			break
		}
		var lnums []string
		for _, n := range limit(ls, 10) {
			lnums = append(lnums, fmt.Sprintf("L%d", n))
		}
		more := ""
		if len(ls) > 10 {
			more = fmt.Sprintf(", +%d more", len(ls)-10)
		}
		out = append(out, fmt.Sprintf("%s: %d %s (%s%s)", file, len(ls), plural(len(ls), "match", "matches"), strings.Join(lnums, ", "), more))
		shown += len(ls)
	}
	return strings.Join(out, "\n")
}

func fileExists(p string) bool { _, err := os.Stat(p); return err == nil }

// ---- find ------------------------------------------------------------------

func HandleFind(args []string, ultra bool) Result {
	start := time.Now()
	var lines []string
	var raw string
	exit := 0

	// The Windows `find` is a substring search with totally different args, so fall back to
	// a portable walker there (and whenever an arg our walker doesn't implement shows up).
	if runtime.GOOS == "windows" || hasUnsupportedFindArg(args) {
		dir, pattern := parseFindArgs(args)
		lines = walkFiles(dir, pattern)
		raw = strings.Join(lines, "\n")
	} else {
		r := run.Run("find", args)
		raw = combined(r)
		exit = r.ExitCode
		for _, l := range strings.Split(util.StripAnsi(raw), "\n") {
			if strings.TrimSpace(l) != "" {
				lines = append(lines, l)
			}
		}
	}

	var filtered string
	switch {
	case ultra:
		filtered = fmt.Sprintf("%d files", len(lines))
	case len(lines) > 50:
		filtered = fmt.Sprintf("%d files found\n%s\n[+%d more]", len(lines), strings.Join(limit(lines, 50), "\n"), len(lines)-50)
	case len(lines) == 0:
		filtered = "0 files"
	default:
		filtered = strings.Join(lines, "\n")
	}
	return Result{Filtered: filtered, Exit: exit, Raw: raw, CmdType: "find", ExecMs: since(start)}
}

func hasUnsupportedFindArg(args []string) bool {
	supported := map[string]bool{"-name": true, "-iname": true}
	for _, a := range args {
		if strings.HasPrefix(a, "-") && !supported[a] {
			return true
		}
	}
	return false
}

func parseFindArgs(args []string) (string, *regexp.Regexp) {
	dir := "."
	var pattern *regexp.Regexp
	for i := 0; i < len(args); i++ {
		a := args[i]
		switch {
		case a == "-name" || a == "-iname":
			glob := ""
			if i+1 < len(args) {
				i++
				glob = args[i]
			}
			pattern = globToRegex(glob, a == "-iname")
		case !strings.HasPrefix(a, "-") && i == 0:
			dir = a
		}
	}
	return dir, pattern
}

func globToRegex(glob string, ci bool) *regexp.Regexp {
	var b strings.Builder
	b.WriteString("^")
	for _, ch := range glob {
		switch {
		case ch == '*':
			b.WriteString(".*")
		case ch == '?':
			b.WriteString(".")
		case strings.ContainsRune(`.+^${}()|[]\`, ch):
			b.WriteByte('\\')
			b.WriteRune(ch)
		default:
			b.WriteRune(ch)
		}
	}
	b.WriteString("$")
	flags := ""
	if ci {
		flags = "(?i)"
	}
	re, err := regexp.Compile(flags + b.String())
	if err != nil {
		return nil
	}
	return re
}

func walkFiles(dir string, pattern *regexp.Regexp) []string {
	var results []string
	queue := []string{dir}
	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		entries, err := os.ReadDir(cur)
		if err != nil {
			continue
		}
		for _, e := range entries {
			full := filepath.Join(cur, e.Name())
			switch {
			case e.IsDir():
				if e.Name() != "node_modules" && e.Name() != ".git" {
					queue = append(queue, full)
				}
			case e.Type().IsRegular():
				if pattern == nil || pattern.MatchString(e.Name()) {
					results = append(results, full)
				}
			}
		}
	}
	return results
}

// ---- diff ------------------------------------------------------------------

func HandleDiff(args []string, ultra bool) Result {
	r := run.Run("diff", args)
	raw := combined(r)
	lines := strings.Split(util.StripAnsi(raw), "\n")
	added, removed := 0, 0
	for _, line := range lines {
		switch {
		case strings.HasPrefix(line, ">") || (strings.HasPrefix(line, "+") && !strings.HasPrefix(line, "+++")):
			added++
		case strings.HasPrefix(line, "<") || (strings.HasPrefix(line, "-") && !strings.HasPrefix(line, "---")):
			removed++
		}
	}
	filtered := fmt.Sprintf("+%d -%d (%d diff lines)", added, removed, len(lines))
	if ultra {
		filtered = fmt.Sprintf("+%d-%d", added, removed)
	}
	return Result{Filtered: filtered, Exit: r.ExitCode, Raw: raw, CmdType: "diff", ExecMs: r.ExecMs}
}
